### PR TITLE
Add transaction isolation support

### PR DESCRIPTION
**Pull request by ayroblu**
Attempt to solve #581 
Probably needs some extra tests, not sure where the best place to put them is.
Also probably needs to implement for all drivers rather than just half of them

Syntax is `knex.transaction(...).setIsolationLevel('string')`

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -328,6 +328,10 @@ interface DMLOptions {
   includeTriggerModifications?: boolean;
 }
 
+interface TransactionMethods<T> {
+  setIsolationLevel: (value: string) => T;
```
Comment (kibertoad) [inline types/index.d.ts:332]: Would be better to use an enum with all possible values, also runtime code should throw an error if unknown isolation level is passed.

File: test/unit/knex.js [hunk 2]
```diff
@@ -447,6 +447,13 @@ describe('knex', () => {
       return knex.destroy();
     });
 
+    it('supports transaction isolation', async () => {
```
Comment (kibertoad) [inline test/unit/knex.js:450]: Would be better to run it in an integration test to cover all the dbs. See test/integration2 for a reference.
If you have Docker installed, you can use `npm run db:start:no-oracle` to have all the dbs locally.
Also test needs to ensure that isolation level is actually doing something.

File: lib/knex-builder/make-knex.js [hunk 3]
```diff
@@ -141,10 +141,8 @@ function initContext(knexFn) {
         const trx = this.client.transaction(container, config, outerTx);
         return trx;
       } else {
-        return new Promise((resolve, reject) => {
-          const trx = this.client.transaction(resolve, config, outerTx);
-          trx.catch(reject);
```
Comment (kibertoad) [inline lib/knex-builder/make-knex.js:146]: I don't see this catch blocked added elsewhere, was its removal intentional?

Comment (ayroblu) [reply lib/knex-builder/make-knex.js:146]: I don't really get why this Promise is here, when in the other branch it just returns the trx element, so ideally I'd find a way to remove it, based on the tests failing, that might not be possible :/

File: lib/dialects/postgres/transaction.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const Transaction = require('../../execution/transaction');
+
+class Transaction_PG extends Transaction {
+  begin(conn) {
+    if (this.isolationLevel) {
+      return this.query(conn, `BEGIN ISOLATION LEVEL ${this.isolationLevel};`);
+    } else {
```
Comment (kibertoad) [inline lib/dialects/postgres/transaction.js:7]: `else` is redundant here.

File: lib/dialects/sqlite3/transaction.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+const Transaction = require('../../execution/transaction');
+
+class Transaction_Sqlite extends Transaction {
+  begin(conn) {
+    const setIsolationLevel = this.isolationLevel
```
Comment (kibertoad) [inline lib/dialects/sqlite3/transaction.js:5]: Does it have to be different from how it's done in PG, which has no ternary operator and `Promise.resolve`?

File: lib/execution/transaction.js [hunk 6]
```diff
@@ -78,7 +78,10 @@ class Transaction extends EventEmitter {
   }
 
   begin(conn) {
-    return this.query(conn, 'BEGIN;');
+    const setIsolationLevel = this.isolationLevel
+      ? this.query(conn, `SET ISOLATION LEVEL ${this.isolationLevel};`)
```
Comment (kibertoad) [inline lib/execution/transaction.js:82]: Which databases would that work for? I think this will currently affect Oracle and MySQL.

File: lib/execution/transaction.js [hunk 7]
```diff
@@ -93,6 +96,11 @@ class Transaction extends EventEmitter {
     return this.query(conn, `RELEASE SAVEPOINT ${this.txid};`, 1, value);
   }
 
+  setIsolationLevel = (value) => {
+    this.isolationLevel = value;
```
Comment (kibertoad) [inline lib/execution/transaction.js:100]: Would be good to validate value here.

File: lib/execution/transaction.js [hunk 8]
```diff
@@ -180,6 +188,9 @@ class Transaction extends EventEmitter {
           // Directly thrown errors are treated as automatic rollbacks.
           let result;
           try {
+            if (!container) {
```
Comment (kibertoad) [inline lib/execution/transaction.js:191]: Why this change?

File: lib/knex-builder/make-knex.js (hunk 3 above)
Comment (kibertoad) [reply lib/knex-builder/make-knex.js:146]: Saying that our transaction logic is messy and complicated would be an... understatement. There are lots of edge cases there.

File: types/index.d.ts (hunk 1 above)
Comment (ayroblu) [reply types/index.d.ts:332]: Well normally there are 4 levels, mssql has 5 (kinda 6), sqlite has 1, kinda 2, from my testing mysql (5.7) says it has 4 but really only works with 3. So I'm not sure exactly the best interface for it

File: lib/dialects/mssql/transaction.js [hunk 9]
```diff
@@ -95,3 +97,27 @@ module.exports = class Transaction_MSSQL extends Transaction {
...
   }
 };
+
+/** MSSQL is annoying this way */
+function nameToIsolationLevelEnum(level) {
+  if (!level) return;
+  level = level.toUpperCase().replace(' ', '_');
+  const knownEnum = isolationEnum[level];
+  if (!knownEnum) {
+    const keys = Object.keys(isolationEnum).map((key) =>
```
Comment (kibertoad) [inline lib/dialects/mssql/transaction.js:107]: I would suggest calculating this once and just having this as a const.

File: lib/dialects/sqlite3/transaction.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+const Transaction = require('../../execution/transaction');
+
+class Transaction_Sqlite extends Transaction {
+  begin(conn) {
+    // SQLite doesn't really support isolation levels, it is serializable by
```
Comment (kibertoad) [inline lib/dialects/sqlite3/transaction.js:5]: Throwing an error is fine.

File: types/index.d.ts (hunk 1 above)
Comment (kibertoad) [reply types/index.d.ts:332]: Supporting all 5 (or) 6 on validation level is fine, DB will throw an error if it doesn't support something.

**Comment by ayroblu** [comment]
Urg I don't get why it's still failing, it passes fine locally :/
image
And I get unrelated failures when I run the whole thing
image

**Comment by ayroblu** [comment]
Okay, it's timing out on `ALTER DATABASE knex_test SET ALLOW_SNAPSHOT_ISOLATION ON` in CI hmmm

**Comment by ayroblu** [comment]
Well that should cover the main parts, let me know if you'd like to change anything, the original issue suggests a different API and I'm not really familiar enough with all the chaining options in terms of what's best.
Probably I guess you might want some more tests I guess otherwise? I had to change the docker compose file cause I couldn't get the alter table to pass so not sure I can do much about that.

File: lib/execution/transaction.js [hunk 11]
```diff
@@ -78,7 +94,13 @@ class Transaction extends EventEmitter {
   }
 
   begin(conn) {
-    return this.query(conn, 'BEGIN;');
+    const setIsolationLevel = this.isolationLevel
```
Comment (kibertoad) [inline lib/execution/transaction.js:97]: Might be  a good idea to rename this variable so that it doesn't have same name as the class method.

File: lib/execution/transaction.js [hunk 12]
```diff
@@ -93,6 +115,18 @@ class Transaction extends EventEmitter {
     return this.query(conn, `RELEASE SAVEPOINT ${this.txid};`, 1, value);
   }
 
+  setIsolationLevel(isolationLevel) {
```
Comment (kibertoad) [inline lib/execution/transaction.js:118]: `transaction()` method already has optional `config` parameter, why not just add isolation level there?

File: scripts/docker-compose.yml [hunk 13]
```diff
@@ -20,7 +20,7 @@ services:
       - bash
       - -c
       # https://docs.microsoft.com/en-us/sql/relational-databases/logs/control-transaction-durability?view=sql-server-ver15#bkmk_DbControl
-      - 'until /opt/mssql-tools/bin/sqlcmd -S mssql -U sa -P S0meVeryHardPassword -d master -Q "CREATE DATABASE knex_test; ALTER DATABASE knex_test SET DELAYED_DURABILITY = FORCED"; do sleep 5; done'
+      - 'until /opt/mssql-tools/bin/sqlcmd -S mssql -U sa -P S0meVeryHardPassword -d master -Q "CREATE DATABASE knex_test; ALTER DATABASE knex_test SET ALLOW_SNAPSHOT_ISOLATION ON; ALTER DATABASE knex_test SET DELAYED_DURABILITY = FORCED"; do sleep 5; done'
```
Comment (kibertoad) [inline scripts/docker-compose.yml:23]: This can't be applied in runtime using knex.raw?

File: test/integration2/transaction/set-isolation-level.spec.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            return;
+          }
+          const trx = await knex
+            .transaction()
+            .setIsolationLevel('read committed');
+          const result1 = await trx(tableName).select();
+          await knex(tableName).insert({ id: 1, value: 1 });
+          const result2 = await trx(tableName).select();
+          await trx.commit();
+          expect(result1).to.not.equal(result2);
```
Comment (kibertoad) [inline test/integration2/transaction/set-isolation-level.spec.js:42]: Wouldn't this assertion pass even if you didn't use transaction isolation? One of the selects happens before the insert, so it is obvious that results won't match.

File: test/integration2/transaction/set-isolation-level.spec.js [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        it('Expect to avoid read skew when repeatable read (snapshot isolation)', async () => {
+          if (isSQLite(knex)) {
+            return;
+          }
+          // NOTE: for mssql, it requires an alter database call that happens in docker-compose
+          const isolationLevel = isMssql(knex) ? 'snapshot' : 'repeatable read';
+          const trx = await knex
+            .transaction()
+            .setIsolationLevel(isolationLevel);
+          const result1 = await trx(tableName).select().timeout(500);
```
Comment (kibertoad) [inline test/integration2/transaction/set-isolation-level.spec.js:54]: why do we need timeout?

File: types/index.d.ts [hunk 16]
```diff
@@ -349,11 +355,11 @@ interface Knex<TRecord extends {} = any, TResult = unknown[]>
   transaction(
     transactionScope?: null,
     config?: any
-  ): Promise<Knex.Transaction>;
+  ): Promise<Knex.Transaction> & TransactionMethods<Promise<Knex.Transaction>>;
```
Comment (kibertoad) [inline types/index.d.ts:358]: Shouldn't this also be added to TransactionProvider?

**Comment by kibertoad** [comment]
Thank you for the contribution! Could you also please update the documentation? https://github.com/knex/documentation

File: lib/dialects/sqlite3/transaction.js [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+    // default and so we override it to ignore isolation level.
+    // There is a `PRAGMA read_uncommitted = true;`, but that's probably not
+    // what the user wants
+    if (this.isolationLevel) {
+      throw new Error(
```
Comment (kibertoad) [inline lib/dialects/sqlite3/transaction.js:10]: Having slept on this, I think warning that the isolation level will be ignored would be better, since SQLite is often used as a stand-in for a real DB, and having "if (test_env)" conditions in code would be a huge hassle.
I think transaction has reference to the client, so it should be possible to do `this.client.logger.warn`.

File: scripts/docker-compose.yml (hunk 13 above)
Comment (ayroblu) [reply scripts/docker-compose.yml:23]: I tried this, but it threw an error, I guess there's side effects elsewhere
`ALTER DATABASE knex_test SET SNAPSHOT ON WITH ROLLBACK IMMEDIATE`
https://stackoverflow.com/a/1253452/6725059

File: test/integration2/transaction/set-isolation-level.spec.js (hunk 14 above)
Comment (ayroblu) [reply test/integration2/transaction/set-isolation-level.spec.js:42]: Yes that's the point, even though you're using a transaction, you aren't getting the isolation that you might expect when running queries, which is the same as not having a transaction at all.

File: test/integration2/transaction/set-isolation-level.spec.js (hunk 15 above)
Comment (ayroblu) [reply test/integration2/transaction/set-isolation-level.spec.js:54]: You don't, I can remove that, it was just helped my testing

File: types/index.d.ts (hunk 16 above)
Comment (ayroblu) [reply types/index.d.ts:358]: Uhhh I haven't touched the TransactionProvider, so I'm not sure if it's the same code path, but sure, I could add it there

File: lib/execution/transaction.js (hunk 12 above)
Comment (ayroblu) [reply lib/execution/transaction.js:118]: Ummm, I could? It's basically the same, but I've never seen any documentation for the config param and I thought the chaining style was more idiomatic 🤷

Comment (kibertoad) [reply lib/execution/transaction.js:118]: Yeah, I'll add types and documentation for the config param. And currently there is no overload for .transaxtion with just a config param, I think.

**Comment by ayroblu** [comment]
@kibertoad Can you help me with oracle? I'm not sure I understand the error, or how oracle really works, maybe worth skipping?
`Error: Transaction query already complete, run with DEBUG=knex:tx for more info`

Docs for isolation with oracle are here: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/SET-TRANSACTION.html#GUID-F11E1E30-5871-48D1-8266-F80A1DF126A1

File: lib/knex-builder/make-knex.js [hunk 18]
```diff
@@ -123,6 +123,12 @@ function initContext(knexFn) {
     // If container is not provided, returns a promise with a transaction that is resolved
     // when transaction is ready to be used.
     transaction(container, _config) {
+      // Overload support of `transaction(config)`
+      if (!_config && typeof container === 'object') {
```
Comment (kibertoad) [inline lib/knex-builder/make-knex.js:127]: It would be safer to use method `isObject` from `lib/util/is.js`, because null is also an object.

File: lib/dialects/oracledb/transaction.js [hunk 19]
```diff
@@ -2,9 +2,26 @@ const Transaction = require('../../execution/transaction');
...
+  begin(conn) {
+    if (this.isolationLevel) {
+      // There's also a "read only", but that's not really an "isolationLevel"
+      if (!supportedIsolationLevels.includes(this.isolationLevel)) {
+        this.client.logger.warn(
+          'Oracle only supports read committed and serializable transactions, ignoring the isolation level param'
+        );
+      } else {
+        this.client.logger.warn(
+          'Not supporting Oracle transaction isolation right now'
```
Comment (kibertoad) [inline lib/dialects/oracledb/transaction.js:18]:
```suggestion
          'Transaction isolation is not currently supported for Oracle'
```

File: lib/dialects/oracledb/transaction.js [hunk 20]
```diff
@@ -2,9 +2,26 @@ const Transaction = require('../../execution/transaction');
...
+        );
+        // I tried this, but it didn't work
```
Comment (kibertoad) [inline lib/dialects/oracledb/transaction.js:20]: Don't have Oracle instance locally, and debugging with Travis is a pain, so yeah, it's fine to skip Oracle for now.

**Review by kibertoad — APPROVED** [review]

**Comment by clemens-smartparking** [comment]
It's not on npm yet, right? or do I miss something? what is the release process?

**Comment by kibertoad** [comment]
@clemens-smartparking Basically the last two things left before 0.95.0 goes full on live: 
1) @tylerjwatson finishes error reporting improvement he needed to get feature parity with his branch;
2) @tylerjwatson confirms that preview version of a new mssql driver behaves correctly.

In case you don't use mssql or would like to contribute to testing the (considerably improved) mssql driver, you can use `0.95.0-next1` already, it is 99.9% what `0.95.0` final will be.

**Comment by tylerjwatson** [comment]
Many apologies for the delay, we've had production problems that have sopped up all my free time.  I am working through getting this done, but it may be a little while yet.

**Comment by kibertoad** [comment]
@tylerjwatson No worries! You are doing a big favour to the project, and we are grateful that you find it possible to contribute altogether. We'd still like to release late February; if you don't believe it would be feasible for you to finish before that, we might have to just bite the bullet and release what we have, and rely on knex community for testing it.

**Comment by clemens-smartparking** [comment]
Thanks for the clarification and thanks for your work! I planned to test for sqlite and pg. However, there seem to be some TS type problems with objection.js when using 0.95.0-next1. Any easy fix for that?

**Comment by kibertoad** [comment]
@clemens-smartparking Unfortunately, that is not that straightforward. First of all, Objection TS types need to be adjusted, as seen in https://github.com/Vincit/objection.js/pull/1992. Second, looks like Objection is importing some of knex internal files directly, so that will need to be adjusted as well. I'll spend some time this week to fix incompatibilities on Objection side, hopefully some kind of `next` release can be done for Objection as well.
